File: problems/problems_72/solution.ts
function minDistance(word1: string, word2: string): number {
    const m: number = word1.length, n: number = word2.length;
	const dp: number[][] = new Array(m + 1).fill(0).map(() => new Array(n + 1).fill(0));
	for (let i: number = 0; i <= m; i++) {
		dp[i][0] = i;
	}
	for (let j: number = 0; j <= n; j++) {
		dp[0][j] = j;
	}
	for (let i: number = 1; i <= m; i++) {
		for (let j: number = 1; j <= n; j++) {
			if (word1[i - 1] === word2[j - 1]) {
				dp[i][j] = dp[i - 1][j - 1];
			} else {
				dp[i][j] = Math.min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]) + 1;
			}
		}
	}
	return dp[m][n];
};

export function Solve(inputJsonElement: string): any {
	const inputValues: string[] = inputJsonElement.split("\n");
	const word1: string = JSON.parse(inputValues[0]);
	const word2: string = JSON.parse(inputValues[1]);
	return minDistance(word1, word2);
}

File: problems/problems_72/solution.rs
use serde_json::{json, Value};

pub struct Solution;

impl Solution {
    pub fn min_distance(word1: String, word2: String) -> i32 {
		let m: usize = word1.len();
		let n: usize = word2.len();
		let mut dp: Vec<Vec<i32>> = vec![vec![0; n + 1]; m + 1];
		for i in 0..=m {
			dp[i][0] = i as i32;
		}
		for j in 0..=n {
			dp[0][j] = j as i32;
		}
		for i in 1..=m {
			for j in 1..=n {
				if word1.chars().nth(i - 1).unwrap() == word2.chars().nth(j - 1).unwrap() {
					dp[i][j] = dp[i - 1][j - 1];
				} else {
					dp[i][j] = dp[i - 1][j - 1].min(dp[i - 1][j].min(dp[i][j - 1])) + 1;
				}
			}
		}
		dp[m][n]
    }
}

#[cfg(feature = "solution_72")]
pub fn solve(input_string: String) -> Value {
	let input_values: Vec<String> = input_string.split('\n').map(|x| x.to_string()).collect();
	let word1: String = serde_json::from_str(&input_values[0]).expect("Failed to parse input");
	let word2: String = serde_json::from_str(&input_values[1]).expect("Failed to parse input");
	json!(Solution::min_distance(word1, word2))
}
